File: solutions/haskell/space-age/1/src/SpaceAge.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE NumericUnderscores #-}

module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

earthYearInSeconds :: Float
earthYearInSeconds = 31_557_600

yearInSeconds :: Planet -> Float
yearInSeconds planet =
    earthYearInSeconds
    * case planet of
        Mercury ->   0.2408467
        Venus   ->   0.61519726
        Earth   ->   1
        Mars    ->   1.8808158
        Jupiter ->  11.862615
        Saturn  ->  29.447498
        Uranus  ->  84.016846
        Neptune -> 164.79132

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / yearInSeconds planet

File: solutions/c/space-age/1/space_age.c
#include "space_age.h"

float YEAR_FACTORS[] = {
    0.2408467,   // MERCURY
    0.61519726,  // VENUS
    1,           // EARTH
    1.8808158,   // MARS
    11.862615,   // JUPITER
    29.447498,   // SATURN
    84.016846,   // URANUS
    164.79132,   // NEPTUNE
};

float age(planet_t planet, int64_t seconds) {
    return seconds / (31557600 * YEAR_FACTORS[planet]);
}
